aztec/src/main/kotlin/org/wordpress/aztec/AztecText.kt

```diff
@@ -462,29 +455,27 @@ open class AztecText : AppCompatEditText, TextWatcher, UnknownHtmlSpan.OnUnknown
         // https://android-review.googlesource.com/c/platform/frameworks/base/+/634929
         val dynamicLayoutCrashPreventer = InputFilter { source, start, end, dest, dstart, dend ->
             var temp : CharSequence? = null
-            if (!bypassCrashPreventerInputFilter
-                    && dstart == dend && dest.length > dend+1
-                    && source != Constants.NEWLINE_STRING) {
-                // dstart == dend means this is an insertion
-                // avoid handling anything if it's a newline
+            if (!bypassCrashPreventerInputFilter) {
+
                 // if there are any images right after the destination position, hack the text
-                val spans = dest.getSpans(dstart, dend+1, AztecImageSpan::class.java)
+                val spans = dest.getSpans(dend, dend+1, AztecImageSpan::class.java)
                 if (spans.isNotEmpty()) {
+
                     // prevent this filter from running twice when `text.insert()` gets called a few lines below
                     disableCrashPreventerInputFilter()
                     // disable MediaDeleted listener before operating on content
                     disableMediaDeletedListener()
 
-                    // take the source (that is, what is being inserted), and append the Image to it. We will delete
-                    // the original Image later so to not have a duplicate.
-                    // use Spannable to copy / keep the current spans
-                    temp = SpannableStringBuilder(source).append(dest.subSequence(dend, dend+1))
+                    // create a new Spannable to perform the text change here
+                    var newText = SpannableStringBuilder(dest.subSequence(0, dstart))
+                            .append(source.subSequence(start, end))
+                            .append(dest.subSequence(dend, dest.length));
+
+                    // force a history update to ensure the change is recorded
+                    history.beforeTextChanged(this@AztecText)
 
-                    // delete the original AztecImageSpan
-                    text.delete(dend, dend+1)
-                    // now insert both the new insertion _and_ the original AztecImageSpan
-                    text.insert(dend, temp)
-                    temp = "" // discard the original source parameter as an ouput from this InputFilter
+                    // use HTML from the new text to set the state of the editText directly
+                    fromHtml(toFormattedHtml(newText), false)
```

Comment:
Just thinking aloud here - we discussed this of course but I'm not sure about doing such a radical thing for _everyone_ with Android 8.x just as you already rightfully commented on the PR description:

> Also, in that PR, strict criteria must be met before the fix is applied. One such criterion is that the selection must be a caret (i.e. no selection). This means certain edge cases will still encounter the crash.

We may want to narrow down the cases so this doesn't introduce a regression for other users who maybe weren't experiencing the problem so, thinking in that direction it seems to me it'd be nice if we kept the "case detection logic" (such as the big `if` clause that was up there) and add / remove more constraints as we see fit. I'm not really  convinced on one way or the other, so please take this comment just as to keep the conversation going and see what we can get from it rather than an indication / suggestion of  "what seems right" in code 👍 

OTOH it may be the only way to cover "all" the possibilities where the issue may appear. I believe we're still spot on in that Images are the ones internally using `DynamicLayout` so it seems right to do the search for it (as in, the broader case).
I'll do some more smoke tests, another thing that comes to mind is maybe we also need to try and keep the caret position as well? Though we may want to address that as a separate PR if it starts to get convoluted

Reply:
> and add / remove more constraints as we see fit

I totally agree with this. So far, these are the conditions that must be met:

1. Android API version must be either 26 or 27.
2. The crash preventer must not already be "in progress" (i.e. no recursion).
3. At least one character must follow the selection / caret.
4. The character immediately following the selection / caret must have at least one `AztecImageSpan` (i.e. it is an image).

Conditions 2 and 3 are set in the outer-most conditional of the filter. Condition 4 requires querying of the `Spannable`, and thus lives inside that outer conditional. If there is a way to determine whether we are on a version of Oreo which has been patched, we could narrow condition 1 further, but I'm not sure this is currently possible.

If you have other conditions that come to mind, we can add them. Note: I intentionally left out the condition regarding the new line character because this approach was different enough, I thought it best to test the behavior separately. Do you know if the current state reproduces (regresses) the issue that condition was added for?

Reply:
> another thing that comes to mind is maybe we also need to try and keep the caret position as well? Though we may want to address that as a separate PR if it starts to get convoluted

I noticed this on `develop` as well, so I think a separate PR would be better, with an issue to track it.

Reply:
> Note: I intentionally left out the condition regarding the new line character because this approach was different enough, I thought it best to test the behavior separately. Do you know if the current state reproduces (regresses) the issue that condition was added for?

I think you're right - just tested that the conditions this PR here https://github.com/wordpress-mobile/AztecEditor-Android/pull/809 was fixing are not met with this PR anymore and wasn't able to reproduce the crash so, I think we're good 👍

Reply:
hey @mkevins I had to re-introduce the check in f77e5764 (hope you don't mind!) to fix the connected tests that were failing for this PR 👍